File: mlproject/QPlayer.java
import java.util.Map;
import java.util.HashMap;
import java.util.Random;

public class QPlayer{
	int mark,mark2;
	QMap qMap;
	Position lastAction;
	double lastReward=0;
	int[] lastState;
	static double prob = 0.5;
	static final double ALPHA = 0.5;
	static final double GAMMA = 0.9;
	Board board;

	QPlayer(Board board,int mark){
		this.mark = mark;
		this.mark2 = Mark.getOpposite(mark);
		// qMap = new HashMap<String,int[][]>();
		// qMap = new QMapArray(board.getSize());
		qMap = new QMapNeural(board.getSize());
		this.board = board;
	}
	void setQMap(QMap map){
		qMap = map;
	}
	QMap getQMap(){
		return qMap;
	}
	Board getBoard(){
		return board;
	}
	boolean chooseToPlay(){
		return new Random().nextDouble() < prob;
	}

	static void increaseProb(){
		if(prob<1)
			prob+=0.1;
	}

	void play(){
		int[] state = board.getState();
		Position action = new Position();
		Random random = new Random();
		double qMax = -1;
		double[] qValues = qMap.get(state);
		if(qValues==null){
			qValues = new double[board.getSize()*board.getSize()];
			for(int i=0;i<board.getSize();i++)
				for(int j=0;j<board.getSize();j++){
					qValues[i*board.getSize()+j] = findReward(new Position(i,j));
				}
			qMap.update(state,qValues);
		}
		for(int i=0;i<board.getSize();i++){
			for(int j=0;j<board.getSize();j++){
				if(board.isBlank(i,j)){
					if(qValues[i*board.getSize()+j]>qMax){
						qMax = qValues[i*board.getSize()+j];
						action = new Position(i,j);
					}
				}
			}
		}

		if(lastState!=null){
			double[] lastQvals = qMap.get(lastState);
			double lastQval = lastQvals[lastAction.i*board.getSize()+lastAction.j];
			double newQVal = lastQval*(1-ALPHA)+ALPHA*(lastReward+GAMMA*qMax);
			qMap.update(lastState,lastAction,newQVal);
		}
		if(!chooseToPlay()){
			int l = random.nextInt(board.getSize());
			int m = random.nextInt(board.getSize());
			while(!board.isBlank(l,m)){
				l = random.nextInt(board.getSize());
				m = random.nextInt(board.getSize());
			}
			action = new Position(l,m);
		}
		lastReward=findReward(action);
		lastState=state;
		lastAction=action;
		board.play(action.i,action.j);
	}

	double findReward(Position action){
		if(!board.isBlank(action))
			return -0.2;
		if(board.willWin(action)){
			return 0.1;
		}
		board.play(action.i,action.j);
		for(int i=0;i<board.getSize();i++){
			for(int j=0;j<board.getSize();j++){
				if(board.willWin(new Position(i,j))){
					board.undo();
					return -0.1;
				}
			}
		}
		board.undo();
		return 0;
	}
}

File: python/qplayer.py
#!/usr/bin/python
from mark import Mark
from position import Position
from qMapArray import QMapArray
import random
class QPlayer(object):
	"""docstring for QPlayer"""
	prob = 0
	alpha = 0.5
	gamma = 0.9
	def __init__(self,board, mark):
		super(QPlayer, self).__init__()
		self.mark = mark
		self.mark2 = Mark.getOpposite(mark)
		self.lastAction = Position()
		self.lastReward = 0
		self.lastState = []
		self.qMap = QMapArray(board.getSize())
		self.board = board
	def setQMap(selfmap):
		self.qMap = map
	def getQMap(self):
		return self.qMap
	def getBoard(self):
		return self.board
	def chooseToPlay(self):
		return random.random() < self.prob
	@classmethod
	def increaseProb(cls):
		if(cls.prob<1):
			cls.prob=cls.prob+0.1
	def play(self):
		state = self.board.getState()
		action = Position()
		qMax = -1
		qValues=self.qMap.get(state)
		for i in range(0,self.board.getSize()):
			for j in range(0,self.board.getSize()):
				if(self.board.isBlank(Position(i,j))):
					if(qValues[i*self.board.getSize()+j]>qMax):
						qMax = qValues[i*self.board.getSize()+j]
						action = Position(i,j)
		if self.lastState:
			# print "Test"
			lastQvals=self.qMap.get(self.lastState)
			lastQval=lastQvals[self.lastAction.i*self.board.getSize()+self.lastAction.j]
			newQVal=lastQval*(1-QPlayer.alpha)+QPlayer.alpha*(self.lastReward+QPlayer.gamma*qMax)
			self.qMap.update(self.lastState,self.lastAction,newQVal)
			# print self.qMap.get(self.lastState)
		if not self.chooseToPlay():
			# print "test"
			l=int(random.random()*self.board.getSize())
			m=int(random.random()*self.board.getSize())
			while not self.board.isBlank(Position(l,m)):
				l=int(random.random()*self.board.getSize())
				m=int(random.random()*self.board.getSize())
			action=Position(l,m)
		self.lastReward=self.findReward(action)
		# print self.lastReward
		self.lastState=state
		self.lastAction=action
		self.board.play(action.i,action.j)
	def findReward(self,action):
		if not self.board.isBlank(action):
			return -0.2
		if self.board.willWin(action):
			return 0.1
		self.board.play(action.i,action.j)
		for i in range(0,self.board.getSize()):
			for j in range(0,self.board.getSize()):
				if(self.board.willWin(Position(i,j))):
					self.board.undo();
					return -0.1
		self.board.undo()
		return 0